# Day 16: Applying the Euler Method

### &#9989; Write your name here

Today, you will apply the Euler method to a physics problem. To review, the Euler method uses a **known derivative** to iteratively approximate values of an **unknown physical quantity**. 

The steps of the Euler method are:
1. Begin with an unknown function $y(x)$, where we **do** know a formula for the derivative $f(x,y) = \frac{dy}{dx}$ and initial values for $y$ and $x$, which we can call $y_0$ and $x_0$. *Note: the function can use different variables than $y$ and $x$, but they should represent a dependent and independent variable, respectively.*
2. Select a range of equally spaced $x$-values to compute over, starting with $x_0$.
3. Go to the next $x$-value, $x_1$. The step-size, $x_1 - x_0$, we can call $\Delta x$.
4. Compute $y(x_1)$, or just $y_1$ for short, by using the Euler method: $ y_1 = y_0 + f(x_0, y_0)\Delta x$.
5. Repeat steps 3 and 4, using the same step-size ($\Delta x$), to get to the next $x$-value.
6. Continue repeating until you reach the end of your range of $x$-values.

When repeating the steps of the Euler method, we can use a more general formula for computing the next value of $y(x)$:

$$ y_{i+1} = y_i + f(x_i, y_i)\Delta x $$

Below, you can visualize this process and see how the formula is derived:

<img src="https://raw.githubusercontent.com/pattihamerski/PH-36X-Public/3118f62aea9eb3c492b76dc3e7461548d2a20c21/image/euler-diagram-with-triangle.png" alt="" width="600"/>

Back in the Day 5 assignment, you modeled a yarn-ball launcher for a cat's birthday party.

The code below shows an **exact** solution using kinematic equations, and it tracks the heights and times that the ball reaches as it travels to its peak.

<img src="https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExcHZndWpkbnV4cm5yaGE5ZXBiZGM3NXJtNDl0eTZndXp6YTY1djR5MyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/jkAs9C3OoA9Nu/giphy.gif" alt="Kitten playing with a ball of yarn" width="250"/>


In [3]:
# Use this code as a reference only
# This code does NOT use the Euler method
# It is only given to serve as a refresher on using while loops 

# set gravitational constant
g = 9.81

# set initial values for height, velocity, time
y0 = 2
vy0 = 44
t0 = 0

# set time step
dt = 0.1

# create arrays for values of t and y to be stored
t_exact = []
y_exact = []

# set up variables for t and y that begin at initial values
t = t0
y = y0

# keep track of most recent y value
y_prev = y

# run the loop as long as y is increasing
while y >= y_prev:
    # update the most recent y value to continue keeping track
    y_prev = y
    
    # append the next t and y values to the arrays
    t_exact.append(t)
    y_exact.append(y)
    
    # increment t by the time step
    t += dt
    # calculate an exact value for y using a kinematic equation
    y = y0 + vy0 * t - 0.5 * g * t ** 2

# when the loop stops, y has reached its peak
# we can see what the maximum height was
print("maximum height is", y_prev)

maximum height is 100.67375


---

### Part 1: Projectile motion with the Euler method

As we know, cats can be sticklers for correct physics (just ask Schrödinger). In this assignment, you must model the yarn-ball's ascent using derivatives. To do so, keep in mind that position, velocity, and acceleration can be defined with respect to derivatives:

$$v_y = \frac{dy}{dt}$$

$$a_y = \frac{dv_y}{dt}$$

We know that $a_y = -g$, where $g=9.81\text{ m/s}^2$. With this in mind, we can solve for $dv_y$:

$$dv_y = -g\cdot dt$$

This means that over a small time-step $dt$, the the vertical velocity will change by an amount $-g\cdot dt$. This information can be used to increment the value for $v_y$. 

A similar logic can be used to increment $y$.

**&#9989; Task 1.1:** Over each time-step $dt$, how much will the height change? In other words, what is $dy$? Express your answer in terms of $v_y$, $g$, and $dt$.

**/your answer here/**

**&#9989; Task 1.2:** Using the Euler method, compute the values of $v_y$ and $y$ at $t=0.1$ and $t=0.2$. Use the same initial conditions and time-steps as in the provided code.

*Note: These are just the first couple values, so you don't need to use a loop. You will be asked to translate your code into a loop in the next task.*

In [4]:
# your answer here

**&#9989; Task 1.3:** Using the Euler method, approximate the maximum height of the yarn-ball at the top of its trajectory. 
- Do not use kinematic equations.
- Use a time step of $dt = 0.1$.
- Store all computed values for $t$ and $y$ in lists or arrays.
- Print the maximum height of the ball at the end.

*Hint: This task can be done using a similar structure to the example code at the top of the assignment.*

In [11]:
# your answer here

**&#9989; Task 1.4:** Using the exact solution **and** your lists or arrays for $t$ and $y$ from Task 1.3, **create a visualization** of the yarn-ball's height over time.
- Lists of **exact values have already been computed** in the example code at the top of the assignment.
- The values represented in your plot should **cut off near the peak** of the trajectory.

In [ ]:
# your answer here

#### &#128721; **Stop here and check your progress with an instructor**

---

### Part 2: Efficiency and precision of the Euler method

The approximation you made in Part 1 is close, but not quite matching the exact solution. An approximation will never be completely exact, but usually it can be made better.

When using Euler method, the best way to **increase precision** is to **decrease the time-step** (or the step-size of the whatever the independent variable is).

Let's create code we can use to test out different levels of precision.

**&#9989; Task 2.1:** In this task, you will be asked to format your code so it is easier to turn it into a function. Copy and paste your code from Task 1.3 below. Identify where $dt$ is first used or defined in your code.
- Move the definition of $dt$ to the very top of the code cell.
- Re-run your code to ensure it still works. If not, edit further until it works as intended, with $dt$ defined at the very top.
- Change the value of $dt$ and re-run your code. It should still work, though maybe it outputs a slightly different value for the maximum height. If not, edit further until it works for different values of $dt$.

In [ ]:
# copy your code from Task 1.3 here, and then edit if needed


**&#9989; Task 2.2:** Use your code from Task 2.1 to **create a function**. This function should do the following.
* Take `dt` as an argument.
* Use the Euler method with a time-step of `dt` to compute values of `t` and `y` for the ball of yarn as it travels up to its peak.
* Print the maximum height reached.
* Output (i.e., `return`) the lists or arrays of `t` and `y` values.

In [ ]:
# your answer here

#### &#128721; **Stop here and check your progress with an instructor**

**&#9989; Task 2.3:** Use your function to compute arrays of `t` values and `y` values that represent solutions for `dt = 0.1`, `0.25`, `0.5`, and `1`. Store these solutions into variables for later use.

In [27]:
# your answer here

**&#9989; Task 2.4:** Create a visualization with the exact solution and all four approximations on it, and include a legend that clearly demonstrates what each curve represents. Make sure to include axis labels and a title as well to clarify the meaning of your visualization. The figure should convey the heights ($y$-axis) and times ($x$-axis) computed in previous tasks, to compare each model for the ball's ascent. It should include:
* The exact solution given at top of the notebook.
* The Euler approximation solutions from Task 2.3.
* A vertical line at the **time** the ball reaches maximum height in the exact solution.

In [ ]:
# your answer here

The time that the ball achieves maximum height in the exact solution can be found using the code snippet below.

In [23]:
# time of maximum height
t_exact[np.argmax(y_exact)]

4.5

It should take 4.5 seconds. With this in mind, the exact solution for maximum height could be computed using an integral like the one below. 

$$y_\text{max} = 2 + \int_0^{4.5}\frac{dy}{dt}dt$$

**&#9989; Task 2.5:** Discuss the following questions with your group, and in your own words, write at least three sentences about your discussion.
* In the visualization you made above, how do the values `dt` affect what you see?
* How does the integral setup relate to your code in Task 2.2?
* What does `dt` represent physically, in the Euler approximation solutions?
* What does $dt$ represent physically, in the integral?
* A finite sum in an approximation of an integral (which is an infinite sum). How can you rewrite the integral as a finite sum to aid in your explanations to the questions above?

**/discussion notes here/**

#### &#128721; **Stop here and check your progress with an instructor**